src/control/ros/RosTrajectoryExecutor.cpp

```diff
@@ -0,0 +1,290 @@
+  , mInProgress{false}
+{
+  if (!mSkeleton)
+    throw std::invalid_argument("Skeleton is null.");
+
+  if (mTimestep <= 0)
+    throw std::invalid_argument("Timestep must be positive.");
+
+  if (mGoalTimeTolerance <= 0)
+    throw std::invalid_argument("Goal time tolerance must be positive.");
+}
+
+//=============================================================================
+RosTrajectoryExecutor::~RosTrajectoryExecutor()
+{
+  // Do nothing.
+  // TODO: Should we wait for the current trajectory to finish executing?
+}
+
+//=============================================================================
+std::future<void> RosTrajectoryExecutor::execute(
+  trajectory::TrajectoryPtr _traj)
+{
+  static const ::ros::Time invalidTime;
+  return execute(_traj, invalidTime);
+}
+
+//=============================================================================
+std::future<void> RosTrajectoryExecutor::execute(
+  trajectory::TrajectoryPtr _traj, const ::ros::Time& _startTime)
+{
+  using statespace::dart::MetaSkeletonStateSpace;
+
+  if (!_traj)
+    throw std::invalid_argument("Trajectory is null.");
+
+  const auto space = std::dynamic_pointer_cast<
+    MetaSkeletonStateSpace>(_traj->getStateSpace());
+  if (!space)
+  {
+    throw std::invalid_argument(
+      "Trajectory is not in a MetaSkeletonStateSpace.");
+  }
+
+  const auto trajectorySkeleton = space->getMetaSkeleton();
+  for (const auto dof : mSkeleton->getDofs())
+  {
+    const auto idof = trajectorySkeleton->getIndexOf(dof, false);
+    if (idof == dart::dynamics::INVALID_INDEX)
+    {
+      std::stringstream msg;
+      msg << "Trajectory is missing DegreeOfFreedom '" << dof->getName()
+          << "' for Skeleton '" << dof->getSkeleton()->getName() << "'.";
+      throw std::invalid_argument(msg.str());
+    }
+  }
+
+  // Setup the goal properties.
+  // TODO: Also set the goal properties.
+  control_msgs::FollowJointTrajectoryGoal goal;
+  goal.trajectory.header.stamp = _startTime;
+  goal.goal_time_tolerance = ::ros::Duration(mGoalTimeTolerance);
+
+  // Convert the Aikido trajectory into a ROS JointTrajectory.
+  // TODO: Move this into a helper function.
+  util::StepSequence timeSequence{mTimestep, true, 0., _traj->getDuration()};
+  const auto numDofs = mSkeleton->getNumDofs();
+  const auto numWaypoints = timeSequence.getMaxSteps();
+  const auto numDerivatives = std::min<int>(_traj->getNumDerivatives(), 2);
+
+  goal.trajectory.joint_names.reserve(numDofs);
+  for (const auto dof : mSkeleton->getDofs())
+    goal.trajectory.joint_names.emplace_back(dof->getName());
+
+  goal.trajectory.points.reserve(numWaypoints);
+  for (const auto timeFromStart : timeSequence)
+  {
+    const auto timeAbsolute = _traj->getStartTime() + timeFromStart;
+    trajectory_msgs::JointTrajectoryPoint waypoint;
+
+    Eigen::VectorXd tangentVector;
+    auto state = space->createState();
+    _traj->evaluate(timeAbsolute, state);
+    space->logMap(state, tangentVector);
```

Comment:
I don't think so. At least, I can't think of any other joints that admit a *unique* parameterization as a single real number. Anything else is difficult to shoehorn into the `JointTrajectory` message.